github/store/db.go

```diff
@@ -0,0 +1,561 @@
+	pullRequestReviewCommentsCols = "author_association, body, commit_id, created_at, diff_hunk, htmlurl, id, in_reply_to, node_id, original_commit_id, original_position, path, position, pull_request_number, pull_request_review_id, repository_name, repository_owner, updated_at, user_id, user_login"
+)
+
+var tables = []string{
+	"organizations_versioned",
+	"issues_versioned",
+	"issue_comments_versioned",
+	"pull_requests_versioned",
+	"pull_request_reviews_versioned",
+	"pull_request_comments_versioned",
+}
+
+func (s *DB) SetActiveVersion(v string) error {
+	// TODO: for some reason the normal parameter interpolation $1 fails with
+	// pq: got 1 parameters but the statement requires 0
+
+	_, err := s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW organizations AS
+	SELECT %s
+	FROM organizations_versioned WHERE '%s' = ANY(versions)`, organizationsCols, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW organizations: %v", err)
+	}
+
+	_, err = s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW users AS
+	SELECT %s
+	FROM users_versioned WHERE '%s' = ANY(versions)`, usersCols, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW users: %v", err)
+	}
+
+	_, err = s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW repositories AS
+	SELECT %s
+	FROM repositories_versioned WHERE '%s' = ANY(versions)`, repositoriesCols, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW repositories: %v", err)
+	}
+
+	_, err = s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW issues AS
+	SELECT %s
+	FROM issues_versioned WHERE '%s' = ANY(versions)`, issuesCols, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW issues: %v", err)
+	}
+
+	_, err = s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW issue_comments AS
+	SELECT %s
+	FROM issue_comments_versioned WHERE '%s' = ANY(versions)`, issueCommentsCols, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW issue_comments: %v", err)
+	}
+
+	_, err = s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW pull_requests AS
+	SELECT %s
+	FROM pull_requests_versioned WHERE '%s' = ANY(versions)`, pullRequestsCol, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW pull_requests: %v", err)
+	}
+
+	_, err = s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW pull_request_reviews AS
+	SELECT %s
+	FROM pull_request_reviews_versioned WHERE '%s' = ANY(versions)`, pullRequestReviewsCols, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW pull_request_reviews: %v", err)
+	}
+
+	_, err = s.DB.Exec(fmt.Sprintf(`CREATE OR REPLACE VIEW pull_request_comments AS
+	SELECT %s
+	FROM pull_request_comments_versioned WHERE '%s' = ANY(versions)`, pullRequestReviewCommentsCols, v))
+	if err != nil {
+		return fmt.Errorf("failed to create VIEW pull_request_comments: %v", err)
+	}
+
+	return nil
+}
+
+func (s *DB) Cleanup(currentVersion string) error {
```

Comment:
Correct me if I'm wrong, but if I run a sync and write into the db an item such as:
```
sha256     | field_1     | field_2     | versions
abcdef     | 1           | 2           | [1]
```
and then this item never changes anymore, then at some point I'll have something like:
```
sha256     | field_1     | field_2     | versions
abcdef     | 1           | 2           | [1, 2, 3, ..., n]
```
because `versions` is _reset_ only on conflict with `sha256`. Then this statement:
```
DELETE FROM %s WHERE '%s' <> ALL(versions)
```
won't delete that record, which is correct, but all those items will take wasted space if we don't prune the `versions` array.

An idea could be to run this statement first:
```
UPDATE %s SET versions = arr[array_upper(arr, 1)]
```

Reply:
That's what the code is already doing in this method. Line 128. Or did I misunderstand your comment?

Reply:
🤦‍♂ I actually don't know how I misread that part, sorry for the noise 😅

Reply:
It's ok, the PR is huge! :sweat_smile: